core/connection/manager.go

```diff
@@ -137,19 +144,62 @@ func (manager *connectionManager) Connect(consumerID identity.Identity, proposal
 
 	sessionDTO, paymentInfo, err := manager.createSession(connection, dialog, consumerID, proposal)
 	if err != nil {
+		manager.sendSessionStatus(dialog, "", connectivity.StatusSessionEstablishmentFailed, err)
 		return err
 	}
 
 	err = manager.launchPayments(paymentInfo, dialog, consumerID, providerID)
 	if err != nil {
+		manager.sendSessionStatus(dialog, sessionDTO.ID, connectivity.StatusSessionPaymentsFailed, err)
 		return err
 	}
 
+	currentPublicIP := manager.getPublicIP()
+	// Try to establish connection with peer.
 	err = manager.startConnection(connection, consumerID, proposal, params, sessionDTO, stateChannel, statisticsChannel)
-	if err == context.Canceled {
-		return ErrConnectionCancelled
+	if err != nil {
+		if err == context.Canceled {
+			return ErrConnectionCancelled
+		}
+		manager.sendSessionStatus(dialog, sessionDTO.ID, connectivity.StatusConnectionFailed, err)
+		return err
+	}
+
+	manager.checkSessionIP(dialog, sessionDTO.ID, currentPublicIP)
+	return nil
+}
+
+// checkSessionIP checks if ip was changed after connection is established and notify peer.
+func (manager *connectionManager) checkSessionIP(dialog communication.Dialog, sessionID session.ID, currentPublicIP string) {
```

Comment:
I'd call `currentPublicIP` something like `originalPublicIP`. 
It's is not really `current` here, and `if currentPublicIP == newPublicIP {` looks wierd. 
```suggestion
func (manager *connectionManager) checkSessionIP(dialog communication.Dialog, sessionID session.ID, originalPublicIP string) {
```

Reply:
Good point, updated.